lua/entities/gmod_wire_expression2/core/player.lua

```diff
@@ -533,3 +536,75 @@ e2function ranger entity:eyeTraceCursor()
 	ret.RealStartPos = this:GetShootPos()
 	return ret
 end
+
+--[[--------------------------------------------------------------------------------------------]]--
+
+local spawnAlert = {}
+local runBySpawn = 0
+local lastJoined = nil
+
+local leaveAlert = {}
+local runByLeave = 0
+local lastLeft = nil
+
+hook.Add("PlayerInitialSpawn","Exp2RunOnJoin", function(ply)
+	runBySpawn = 1
+    lastJoined = ply
+	for e,_ in pairs(spawnAlert) do
+		if IsValid(e) then
+			e:Execute()
+		else
+			spawnAlert[e] = nil
+		end
+	end
+	runBySpawn = 0
+end)
+
+hook.Add("PlayerDisconnected","Exp2RunOnLeave", function(ply)
+	runByLeave = 1
+    lastLeft = ply
+	for e,_ in pairs(leaveAlert) do
+		if IsValid(e) then
+			e:Execute()
+		else
+			leaveAlert[e] = nil
+		end
+	end
+	runByLeave = 0
+end)
+
+
+__e2setcost(3)
+e2function void runOnJoin(activate)
+	if activate ~= 0 then
+		spawnAlert[self.entity] = true
+	else
+		spawnAlert[self.entity] = nil
+	end
+end
+e2function number joinClk()
+	return runBySpawn
+end
+e2function entity lastJoined()
+	if not IsValid(lastJoined) then return nil end
+	if not lastJoined:IsPlayer() then return nil end
+
+	return lastJoined
+end
+
+e2function void runOnDisconnect(activate)
+	if activate ~= 0 then
+		leaveAlert[self.entity] = true
+	else
+		leaveAlert[self.entity] = nil
+	end
+end
+e2function number disconnectClk()
+	return runByLeave
+end
+e2function entity lastDisconnected()
+	if not IsValid(lastLeft) then return nil end
+	if not lastLeft:IsPlayer() then return nil end
```

Comment:
Well I mean, yeah `entity:IsValid` works out because of code flow, although what do you mean by `strong reasons to prefer the method that is consistently good`? `entity:IsValid` will work consistently with this code, just like `IsValid(entity)` will.